storage_service/common/utils.py

```
@@ -100,8 +118,101 @@ def download_file_stream(filepath, temp_dir=None):
+                EP.eventOutcomeDetailNote(event_outcome_detail_note)
+            )
+        ),
+        EP.linkingAgentIdentifier(
+            EP.linkingAgentIdentifierType(agent_type),
+            EP.linkingAgentIdentifierValue(agent_value),
+        ),
+        version='2.2'
+    )
+    premis_event.set('{'+NSMAP['xsi']+'}schemaLocation', 'info:lc/xmlns/premis-v2 http://www.loc.gov/standards/premis/v2/premis-v2-2.xsd')
+
+    # digiprovMD to wrap PREMIS event
+    digiprov_event = E.digiprovMD(
+        E.mdWrap(
+            E.xmlData(premis_event),
+            MDTYPE="PREMIS:EVENT",
+        ),
+        ID=digiprov_id,
+    )
+    return digiprov_event
+
+
+def mets_ss_agent(xml, digiprov_id, agent_value=None, agent_type='storage service'):
+    """
+    Create and return a PREMIS:AGENT for the SS, if not found in `xml`.
+    """
+    if agent_value == None:
+        agent_value = _storage_service_agent()
+    existing_agent = xml.xpath(".//mets:agentIdentifier[mets:agentIdentifierType='{}' and mets:agentIdentifierValue='{}']".format(agent_type, agent_value), namespaces=NSMAP)
+    if existing_agent:
+        return None
+    digiprov_agent = E.digiprovMD(
+        E.mdWrap(
+            E.xmlData(
+                E.agent(
+                    E.agentIdentifier(
+                        E.agentIdentifierType(agent_type),
+                        E.agentIdentifierValue(agent_value),
+                    ),
+                    E.agentName('Archivematica Storage Service'),
+                    E.agentType('software'),
+                )
+            ),
+            MDTYPE='PREMIS:AGENT',
+        ),
+        ID=digiprov_id,
+    )
+    return digiprov_agent
+
+
 ############ OTHER ############
 
+def generate_checksum(file_path, checksum_type='md5'):
+    """
+    Returns checksum object for `file_path` using `checksum_type`.
+    """
+    try:
+        checksum = hashlib.new(checksum_type)
+    except ValueError:  # Invalid checksum_type
```

Maybe it's worth leaving the raise in this case - silently returning a different checksum type could lead to weird hard-to-debug issues later.